File: crc_calculation.c
#include <stdio.h>
#include <stdint.h>

uint16_t calculate_crc(uint8_t *data, uint8_t length) {
    uint16_t crc_register = 0xFFFF;

    for (uint8_t i = 0; i < length; i++) {
        crc_register ^= data[i];

        for (uint8_t j = 0; j < 8; j++) {
            if (crc_register & 0x0001) {
                crc_register = (crc_register >> 1) ^ 0xA001;
            } else {
                crc_register >>= 1;
            }
        }
    }

    return crc_register & 0xFFFF;
}

int main() {
    uint8_t data[] = {0x01, 0x03, 0x04, 0x01, 0xF4, 0x05, 0xDC};
    uint8_t length = sizeof(data) / sizeof(data[0]);

    uint16_t crc = calculate_crc(data, length);
    uint8_t crc_high = (crc >> 8) & 0xFF;
    uint8_t crc_low = crc & 0xFF;

    printf("CRC High: 0x%02X\n", crc_high);
    printf("CRC Low: 0x%02X\n", crc_low);

    return 0;
}

File: crc_calculation.py
def calculate_crc(data):
    crc_register = 0xFFFF

    for byte in data:
        crc_register ^= byte

        for _ in range(8):
            if crc_register & 0x0001:
                crc_register = (crc_register >> 1) ^ 0xA001
            else:
                crc_register >>= 1

    return crc_register & 0xFFFF

# Example usage

#Example 1
data = [0x01, 0x03, 0x00, 0x02, 0x00, 0x02]

#Example 2
#data = [0x01, 0x06, 0x00, 0x00, 0x09, 0x60]

#Example 3
#data = [0x01, 0x10, 0x00, 0x00, 0x00, 0x02, 0x04, 0x09, 0x60, 0x05, 0xDC]

crc = calculate_crc(data)
crc_swapped = ((crc & 0xFF) << 8) | ((crc >> 8) & 0xFF)

print(f"CRC: {crc_swapped:04X}")

